File: lib/algorithms/sort/merge_sort.rb
def merge_sort(arr)

  return arr if arr.length <= 1

  mid = arr.length/2
  left = merge_sort(arr[0...mid])
  right = merge_sort(arr[mid..-1])

  merge(left, right)
end

def merge(left, right)
  answer = []
  until left.empty? || right.empty?
    if left.first > right.first
      answer << right.shift
    else
      answer << left.shift
    end
  end
    
  answer + left + right
end

File: lib/algorithms/sort/merge_sort.js
function mergeSort(array) {

    if (array.length <= 1) {
        return array;
    }
    
    var mid = Math.floor(array.length/2);
    var left = mergeSort(array.slice(0,mid));
    var right = mergeSort(array.slice(mid));
    
    return merge(left, right);
}

function merge(left, right) {
    var answer = [];

    while(left.length > 0 && right.length > 0) {
        if(left[0] < right[0]) {
            answer.push(left.shift());
        } else {
            answer.push(right.shift());
        }
    }
    
    if (left.length) {
        while(left.length >= 1) {
            answer.push(left.shift());
        }
    }
    
    if(right.length) {
        while(right.length >= 1) {
            answer.push(right.shift());
        }
    }
    
    return answer;
}

mergeSort([4,3,2,1]);